### PR TITLE
Adding options for NodeJs ClientRequest

### PR DESCRIPTION
**Pull request by dragoscirjan**

### CODE REVIEW

**Comment by Acconut** [comment]
Thank you for the PR but I don't see this as a good API design. Instead of passing the request options through the tus.Upload class, one could implement a child class of NodeHttpStack which sets the custom request directly. This custom NodeHttpStack can then be used by the `httpStack` option.

Alternatively, the default NodeHttpStack could have an additional method to set Node.js specific request options which are used for every request.

What do you think?

**Comment by dragoscirjan** [comment]
I just remembered why I chose to pass the settings through `tus.Upload` and not through `HttpStack`. None of the `HttpStack` classes are visible outside of the package, and it would be redundant to copy the code outside of the project, just to extend.

I get what you were saying, though, so here is a new approach. This approach will export two `HttpStack` classes though. One which will allow passing options to `http(s).request` and another one which will force acceptance of insecure certificates no matter what.

I don't know whether I should update anything in the TypeScript definitions as well or not.

File: .travis.yml [hunk 1]
```diff
@@ -3,8 +3,8 @@ dist: xenial
 language: node_js
 jobs:
   include:
-  - env: TEST_TARGET=browser
-    node_js: 12
+  # - env: TEST_TARGET=browser
```
Comment (Acconut) [inline .travis.yml:6]: Is there a reason for disabling the browser tests?

File: lib/node/httpStack.js [hunk 2]
```diff
@@ -5,22 +5,44 @@ import {Readable, Transform} from "stream";
...
+  }
+
   createRequest(method, url) {
-    return new Request(method, url);
+    return new Request(method, url, this._requestOptions);
   }
 
   getName() {
     return "NodeHttpStack";
+    // return this.constructor.name;
```
Comment (Acconut) [inline lib/node/httpStack.js:18]: Please remove this unused line.

File: lib/node/httpStack.js [hunk 3]
```diff
@@ -5,22 +5,44 @@ import {Readable, Transform} from "stream";
...
+  }
+}
+
+/**
+ * A NodeHttpStack which will allow developer to customer `http(s).request()`
+ * options, yet force it to accept insecure certificates as well.
+ */
+export class InsecureNodeHttpStack extends NodeHttpStack {
```
Comment (Acconut) [inline lib/node/httpStack.js:26]: Frankly, I don't think we need to export an InsecureNodeHttpStack if the users can simply pass their own options to NodeHttpStack. That just increases the API surface with little benefit IMO.

File: lib/node/httpStack.js [hunk 4]
```diff
@@ -47,10 +69,17 @@ class Request {
...
+        ...this._headers
+      };
       if (body && body.size) options.headers["Content-Length"] = body.size;
+      options = {
+        ...this._requestOptions,
+        ...options
+      };
 
-      let req = this._request = (options.protocol !== "https:") ? http.request(options) : https.request(options);
+      let req = (this._request = options.protocol !== "https:" ? http.request(options) : https.request(options));
```
Comment (Acconut) [inline lib/node/httpStack.js:82]: Can you explain why the change of these parenthesis is about?

File: lib/node/index.js [hunk 5]
```diff
@@ -42,5 +42,7 @@ export {
   // Make FileUrlStorage module available as it will not be set by default.
   FileUrlStorage,
   canStoreURLs,
-  enableDebugLog
+  enableDebugLog,
+  HttpStack,
```
Comment (Acconut) [inline lib/node/index.js:46]: :+1: Can you also add an export for the browser's HttpStack?

File: lib/upload.js [hunk 6]
```diff
@@ -1,9 +1,11 @@
 /* global window */
-import DetailedError from "./error";
-import uuid from "./uuid";
 import { Base64 } from "js-base64";
+
```
Comment (Acconut) [inline lib/upload.js:3]: Can you please remove the empty lines?

File: test/spec/test-node-specific.js [hunk 7]
```diff
@@ -39,7 +39,7 @@ describe("tus", function () {
       expect(err.message).toBe("tus: cannot automatically derive upload's size from input and must be specified manually using the `uploadSize` option");
     });
 
-    it("should reject streams without specifing the chunkSize", async function () {
+    it("should reject streams without specifying the chunkSize", async function () {
```
Comment (Acconut) [inline test/spec/test-node-specific.js:42]: Thank you very much :sweat_smile:

File: test/spec/test-node-specific.js [hunk 8]
```diff
@@ -186,6 +186,13 @@ describe("tus", function () {
       await assertUrlStorage(storage);
     });
   });
+
+  describe("#InsecureNodeHttpStack", function () {
```
Comment (Acconut) [inline test/spec/test-node-specific.js:190]: Speaking of tests, do you know if it's possible to a test whether the options are properly passed to the request object?

File: .travis.yml (hunk 1 above)
Comment (dragoscirjan) [reply .travis.yml:6]: yes. I don't have an account & my tests would fail. will enable them back :)

File: lib/node/httpStack.js (hunk 2 above)
Comment (dragoscirjan) [reply lib/node/httpStack.js:18]: wish I could've used that approach, however, I'm not sure if it's a TS feature or an ES feature. We can leave a string for now.

File: lib/node/httpStack.js (hunk 3 above)
Comment (dragoscirjan) [reply lib/node/httpStack.js:26]: OK, I can leave with this :)

File: lib/node/httpStack.js (hunk 4 above)
Comment (dragoscirjan) [reply lib/node/httpStack.js:82]: I have no clue what that was :)) reverted back

File: lib/upload.js (hunk 6 above)
Comment (dragoscirjan) [reply lib/upload.js:3]: fixed

File: test/spec/test-node-specific.js (hunk 8 above)
Comment (dragoscirjan) [reply test/spec/test-node-specific.js:190]: unfortunately no. most of the properties added to nodejs's `Request` are stored in private varibales

File: lib/node/index.js (hunk 5 above)
Comment (dragoscirjan) [reply lib/node/index.js:46]: Don't merge just yet. I'll try add a browser's HttpStack also, but it may take a little more time.

**Comment by dragoscirjan** [comment]
Most comments are addressed. Give me some time to fix the HttpStack for browser as well :)

**Comment by dragoscirjan** [comment]
OK. I've updated everything, including browser's HttpStack, but unfortunately, my browser tests fail; they failed even before I started changing the code

**Comment by dragoscirjan** [comment]
Ha! They fail on my Travis account. The PR passed :)

**Comment by dragoscirjan** [comment]
One more question. From the same discussion regarding exposing browser.HttpStack, I would expose Request and Response classes as well. There may be people who need them when extending HttpStack. What do you say?
